stacks-signer/src/runloop.rs

```diff
@@ -164,45 +178,159 @@ impl<C: Coordinator> RunLoop<C> {
+                    "Received a block proposal that was rejected by the stacks node: {:?}",
+                    block_validate_reject
+                );
+                // TODO: submit a rejection response to the .signers contract for miners
+                // to observe so they know to ignore it and to prove signers are doing work
+                let block_rejection = BlockRejection {
+                    block: block_validate_reject.block,
+                    reason: block_validate_reject.reason,
+                    reason_code: RejectCode::ValidationFailed(block_validate_reject.reason_code),
+                };
+                let message =
+                    SignerMessage::BlockResponse(BlockResponse::Rejected(block_rejection));
+                if let Err(e) = self
+                    .stackerdb
+                    .send_message_with_retry(self.signing_round.signer_id, message)
+                {
+                    warn!(
+                        "Failed to send block rejection response to stacker-db: {:?}",
+                        e
+                    );
+                }
+            }
+        }
+    }
+
+    // Handle the stackerdb chunk event as a signer message
+    fn handle_stackerdb_chunk_event_signers(
         &mut self,
-        event: &StackerDBChunksEvent,
-    ) -> (Vec<Packet>, Vec<OperationResult>) {
-        // Determine the current coordinator id and public key for verification
+        stackerdb_chunk_event: StackerDBChunksEvent,
+        res: Sender<Vec<OperationResult>>,
+    ) {
         let (_coordinator_id, coordinator_public_key) =
             calculate_coordinator(&self.signing_round.public_keys);
-        // Filter out invalid messages
-        let inbound_messages: Vec<Packet> = event
+
+        let inbound_messages: Vec<Packet> = stackerdb_chunk_event
             .modified_slots
             .iter()
             .filter_map(|chunk| {
-                let packet = bincode::deserialize::<Packet>(&chunk.data).ok()?;
-                if packet.verify(&self.signing_round.public_keys, coordinator_public_key) {
+                // We only care about verified wsts packets. Ignore anything else
+                let signer_message = bincode::deserialize::<SignerMessage>(&chunk.data).ok()?;
+                let packet = match signer_message {
+                    SignerMessage::Packet(packet) => packet,
+                    _ => return None, // This is a message for miners to observe. Ignore it.
+                };
+                if packet.verify(&self.signing_round.public_keys, &coordinator_public_key) {
+                    debug!("Verified wsts packet: {:?}", &packet);
                     Some(packet)
                 } else {
                     None
                 }
             })
             .collect();
+
         // First process all messages as a signer
-        let mut outbound_messages = self
+        // TODO: deserialize the packet into a block and verify its contents
```

Comment:
Presumably, there's also a dependency between validating the block and signing it, right?  What happens if your signer gets a "sign block" request before it has even obtained the block to validate?  Where is this captured in the run loop?

Reply:
I will change it around. i.e. as soon as a block arrives from the miner. Trigger a signing round. and then have every signer receiving the sign request pass it to the endpoint to be validated before continuing the signing round. however, this will be heavily dependent on Joey's changes to the sign logic to enable me to deserialize it. I would leave it to a sep PR.

Reply:
Can you add a ticket for this before we merge this?